### DISCUSSION TITLE
code inside app.frame(endpoint, code) run multiple time

### DISCUSSION THREAD

**Discussion by onchainyaotoshi** [wevm/frog-archived#71] [Q&A] [answered]
![Screenshot_58](https://github.com/wevm/frog/assets/142854977/b69762bb-b2e1-443c-8831-5e4ef72e3115)

When I click the "Let's go" button, why does it run three times? Is this intended?

![Screenshot_59](https://github.com/wevm/frog/assets/142854977/0bb41ca0-c5c4-4bf8-a09d-8dde647a313e)

Here's the let's go code

```
import { Frog } from 'frog'

import Poll from '../controllers/poll'
import Create from '../controllers/poll-create'
import PlaceHolder from '../controllers/placeholder'
// @ts-ignore
import User from '../models/user'

import {app as CreateHandler} from './pool-create'
import { neynar } from 'frog/hubs'

// export const app = new Frog({
//   hub: neynar({ apiKey: process.env.NEYNAR_API_KEY || '' }),
//   verify: 'silent'
// })

export const app = new Frog()

app.frame('/', async (c) => {
  const {buttonValue,status, frameData} = c;
  const fid = frameData?.fid;
  await User.createIfNotExists(fid);
  console.log('pool.tsx', status, buttonValue);
  if(buttonValue == 'poll-create'){
    return Create(c, '/create');
  }else if(buttonValue == 'poll-list'){
    return PlaceHolder(c);
  }else{
    return Poll(c);
  }
})

app.route('/create', CreateHandler);
```

as you can see, it's not advisable to include database operations inside app.frame if it executes multiple times. Are there any workarounds, or am I approaching this the wrong way?

**Comment by onchainyaotoshi** [comment, marked as answer]
i guess i find the answer here:

https://frog.fm/faq#why-is-my-frame-handler-being-invoked-twice

**Reply by jxom** [reply]
Thanks. I made it more obvious by adding a [dedicated "Render Cycles" section](https://frog.fm/concepts/render-cycles).
